src/components/NewBranchDialog.tsx

```diff
@@ -0,0 +1,440 @@
+import * as React from 'react';
+import { classes } from 'typestyle';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import Dialog from '@material-ui/core/Dialog';
+import DialogActions from '@material-ui/core/DialogActions';
+import ClearIcon from '@material-ui/icons/Clear';
+import { showErrorMessage } from '@jupyterlab/apputils';
+import { Git, IGitExtension } from '../tokens';
+import {
+  actionsWrapperClass,
+  activeListItemClass,
+  branchDialogClass,
+  buttonClass,
+  cancelButtonClass,
+  closeButtonClass,
+  contentWrapperClass,
+  createButtonClass,
+  filterClass,
+  filterClearClass,
+  filterInputClass,
+  filterWrapperClass,
+  listItemBoldTitleClass,
+  listItemClass,
+  listItemContentClass,
+  listItemDescClass,
+  listItemIconClass,
+  listItemTitleClass,
+  listWrapperClass,
+  nameInputClass,
+  titleClass,
+  titleWrapperClass
+} from '../style/NewBranchDialog';
+
+const BRANCH_DESC = {
+  current:
+    'The current branch. Pick this if you want to build on work done in this branch.',
+  default:
+    'The default branch. Pick this if you want to start fresh from the default branch.'
+};
+
+/**
+ * Interface describing component properties.
+ */
+export interface INewBranchDialogProps {
+  /**
+   * Git extension data model.
+   */
+  model: IGitExtension;
+
+  /**
+   * Boolean indicating whether to show the dialog.
+   */
+  open: boolean;
+
+  /**
+   * Callback to invoke upon closing the dialog.
+   */
+  onClose: () => void;
+}
+
+/**
+ * Interface describing component state.
+ */
+export interface INewBranchDialogState {
+  /**
+   * Branch name.
+   */
+  name: string;
+
+  /**
+   * Base branch.
+   */
+  base: string;
+
+  /**
+   * Menu filter.
+   */
+  filter: string;
+
+  /**
+   * Current branch name.
+   */
+  current: string;
+
+  /**
+   * Current list of branches.
+   */
+  branches: Git.IBranch[];
+}
+
+/**
+ * React component for rendering a dialog to create a new branch.
+ */
+export class NewBranchDialog extends React.Component<
+  INewBranchDialogProps,
+  INewBranchDialogState
+> {
+  /**
+   * Returns a React component for rendering a branch menu.
+   *
+   * @param props - component properties
+   * @returns React component
+   */
+  constructor(props: INewBranchDialogProps) {
+    super(props);
+
+    const repo = this.props.model.pathRepository;
+    this._addListeners();
```

Comment:
Should be moved in `ComponentWillMount` to avoid state update before rendering

Reply:
Changed. Lifecycle method: `componentDidMount`.